railties/lib/rails/commands/notes/notes_command.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require "rails/source_annotation_extractor"
+
+module Rails
+  module Command
+    class NotesCommand < Base # :nodoc:
+      class_option :annotations, aliases: "-a", desc: "Filter by specific annotations, e.g. Foobar TODO", type: :array, default: %w(OPTIMIZE FIXME TODO)
+
+      def perform(*)
+        require_application_and_environment!
+
+        display_deprecation_warning
+        display_annotations
+      end
+
+      private
+        def display_annotations
+          annotations = options[:annotations]
+          tag = (annotations.length > 1)
+
+          Rails::SourceAnnotationExtractor.enumerate annotations.join("|"), tag: tag, dirs: directories
```

Comment:
@kaspth @rafaelfranca I realized I can't use `say` here or it ends up printing the result of that method which is an array of all the files on top of the desired output which is coming from a bunch of `puts` inside that module. This is not optimal as I'm outputting from a bunch of different places.

I really feel like we should only allow one way to output from a `Rails::Command`. Having this mix of ways to output could end up biting us later if we wanted to output the result in another place than `stdout`. It also bypasses the formatting options coming from Thor's `say`.

I don't think it's worth bloating this PR more than it already is but do you think it would be worth having a follow up PR making `enumerate` return a string we can output with `say` here or are we ok with having outputs coming from both `puts` and `say` across different modules?

Reply:
I think having `puts` littered throughout a helper class is because Rake might be cumbersome to communicate back and forth with (or it just doesn't feel nice to do so).



I like to think of commands as similar to controllers. Where controllers are traffic cops for HTTP params and how to respond over HTTP (or whatever), commands are traffic cops for IO in/out and more in that CLI vein. The latest exploration of that was in the server command, where we moved some printing out of Rails::Server and into the command.

> I really feel like we should only allow one way to output from a Rails::Command.

Perhaps that might be too restrictive. Perhaps there's a worthy experiment in seeing what executing a command within a context where `puts` and `print` are overriden, so we can capture whatever a helper object will output. Otherwise we might have to pass an IO object around everywhere.

I find `say` annoying enough as is. It's too close to `puts` to seem worth writing it, so perhaps we should rename it to something more context appropriate.

There could perhaps also be something about doing `ActiveSupport::Notifications` so you could do log subscribing on commands.

> I don't think it's worth bloating this PR more than it already is but do you think it would be worth having a follow up PR making enumerate return a string we can output with say here or are we ok with having outputs coming from both puts and say across different modules?

Perhaps. There's definitely something about how it's best to communicate between a command and its collaborator objects that could be teased out here.